### PR TITLE
Inject DAPR env vars in usr container

### PR DESCRIPTION
**Pull request by mukundansundar**
# Description
Inject `DAPR_HTTP_PORT` and `DAPR_GRPC_PORT` to user containers.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #1637 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Unit tests passing
* [x] End-to-end tests passing
* [x] Extended the documentation / Created PR in the https://github.com/dapr/docs/ repo: https://github.com/dapr/docs/pull/666
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: 
* [x] Provided sample for the feature / Created PR in the https://github.com/dapr/samples/ repo: https://github.com/dapr/samples/pull/250

### CODE REVIEW

File: pkg/injector/pod_patch.go [hunk 1]
```diff
@@ -56,7 +56,9 @@ const (
 	apiAddress                        = "dapr-api"
 	placementService                  = "dapr-placement"
 	sentryService                     = "dapr-sentry"
+	userContainerDaprHTTPPortName     = "DAPR_HTTP_PORT"
 	sidecarHTTPPortName               = "dapr-http"
+	userContainerDaprGRPCPortName     = "DAPR_GRPC_PORT"
```
Comment (youngbupark) [inline pkg/injector/pod_patch.go:61]: nit: can you move this to below 59 ? If possible, please group them.

File: pkg/injector/pod_patch.go [hunk 2]
```diff
@@ -142,10 +146,57 @@ func (i *injector) getPodPatchOperations(ar *v1beta1.AdmissionReview,
 			Value: value,
 		},
 	)
+	patchOps = append(patchOps, envPatchOps...)
 
 	return patchOps, nil
 }
 
+func setDaprEnvVars(list []corev1.Container, basePath string) []PatchOperation {
```
Comment (youngbupark) [inline pkg/injector/pod_patch.go:154]: Can we rename the function name and argument naming? 

e.g.
```
setDaprEnvVarsToContainers(containers []corev1.Container)
```

File: pkg/injector/pod_patch.go [hunk 3]
```diff
@@ -124,12 +126,14 @@ func (i *injector) getPodPatchOperations(ar *v1beta1.AdmissionReview,
...
 
 	patchOps := []PatchOperation{}
+	envPatchOps := []PatchOperation{}
 	var path string
 	var value interface{}
 	if len(pod.Spec.Containers) == 0 {
 		path = "/spec/containers"
 		value = []corev1.Container{*sidecarContainer}
 	} else {
+		envPatchOps = setDaprEnvVars(pod.Spec.Containers, "/spec/containers")
```
Comment (youngbupark) [inline pkg/injector/pod_patch.go:136]: I would define const for `/spec/containers` which is used in multiple places. then I will not pass this argument.

File: pkg/injector/pod_patch.go [hunk 4]
```diff
@@ -142,10 +146,57 @@ func (i *injector) getPodPatchOperations(ar *v1beta1.AdmissionReview,
...
+
+			if !hasKey {
+				value = addEnv
+				if !first {
+					path += "/-"
+				} else {
+					first = false
+					value = []corev1.EnvVar{addEnv}
+				}
+				log.Infof("env var added to container %s. %s=%s", container.Name, addEnv.Name, addEnv.Value)
```
Comment (youngbupark) [inline pkg/injector/pod_patch.go:188]: this will be very talkative. please use `log.DebugF(..)`

File: pkg/injector/pod_patch_test.go [hunk 5]
```diff
@@ -113,3 +115,101 @@ func TestGetSideCarContainer(t *testing.T) {
 	assert.Equal(t, "secret", container.Env[2].ValueFrom.SecretKeyRef.Name)
 	assert.EqualValues(t, expectedArgs, container.Args)
 }
+
+func TestSetDaprEnvVars(t *testing.T) {
+	mockContainer := corev1.Container{}
```
Comment (youngbupark) [inline pkg/injector/pod_patch_test.go:120]: Please refactor this test code by using [subtests](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks) and [TableDriventests](https://github.com/golang/go/wiki/TableDrivenTests#example-of-a-table-driven-test).

File: pkg/injector/pod_patch.go [hunk 6]
```diff
@@ -142,10 +146,57 @@ func (i *injector) getPodPatchOperations(ar *v1beta1.AdmissionReview,
...
+	var value interface{}
+	for i, container := range list {
+		first := len(container.Env) == 0
+		for _, addEnv := range portEnv {
+			path := fmt.Sprintf("%s/%d/env", basePath, i)
+			hasKey := false
+			for _, actual := range container.Env {
+				if actual.Name == addEnv.Name {
+					hasKey = true
+					log.Infof("env var already set by user in container %s. %s=%s", container.Name, actual.Name, actual.Value)
```
Comment (youngbupark) [inline pkg/injector/pod_patch.go:175]: please use `log.DebugF`

File: pkg/injector/pod_patch.go [hunk 7]
```diff
@@ -142,10 +146,57 @@ func (i *injector) getPodPatchOperations(ar *v1beta1.AdmissionReview,
...
+			Name:  userContainerDaprHTTPPortName,
+			Value: strconv.Itoa(sidecarHTTPPort),
+		},
+		{
+			Name:  userContainerDaprGRPCPortName,
+			Value: strconv.Itoa(sidecarAPIGRPCPort),
+		},
+	}
+	var value interface{}
+	for i, container := range list {
```
Comment (youngbupark) [inline pkg/injector/pod_patch.go:167]: What happens if pods have multiple containers ? e.g. istio envoy or the other sidecars are already injected ?

File: pkg/injector/pod_patch.go [hunk 8]
```diff
@@ -142,10 +146,57 @@ func (i *injector) getPodPatchOperations(ar *v1beta1.AdmissionReview,
...
+		{
+			Name:  userContainerDaprHTTPPortName,
+			Value: strconv.Itoa(sidecarHTTPPort),
+		},
+		{
+			Name:  userContainerDaprGRPCPortName,
+			Value: strconv.Itoa(sidecarAPIGRPCPort),
+		},
+	}
+	var value interface{}
```
Comment (youngbupark) [inline pkg/injector/pod_patch.go:166]: why do we need to define var here? this var is only used in the scope of line 181.

File: pkg/injector/pod_patch.go [hunk 9]
```diff
@@ -142,10 +146,57 @@ func (i *injector) getPodPatchOperations(ar *v1beta1.AdmissionReview,
...
+	for i, container := range list {
+		first := len(container.Env) == 0
+		for _, addEnv := range portEnv {
+			path := fmt.Sprintf("%s/%d/env", basePath, i)
```
Comment (youngbupark) [inline pkg/injector/pod_patch.go:170]: For readability, I would define new function and move the codes from line 170 to 193 to new function. 

e.g.
```
getEnvPatchOperation(envs *container.Envs, addVar coverv1.EnvVar) []PatchOperation {
line 170-193
}
```

then you can remove hasKey var and its if-statement and make this code cleaner and more readable.

File: pkg/injector/pod_patch.go [hunk 10]
```diff
@@ -142,10 +146,57 @@ func (i *injector) getPodPatchOperations(ar *v1beta1.AdmissionReview,
...
+				if actual.Name == addEnv.Name {
+					hasKey = true
+					log.Infof("env var already set by user in container %s. %s=%s", container.Name, actual.Name, actual.Value)
+					break
+				}
+			}
+
+			if !hasKey {
+				value = addEnv
+				if !first {
```
Comment (youngbupark) [inline pkg/injector/pod_patch.go:182]: I would add comments why do we care the first envvar object specially.

**Comment by youngbupark** [comment]
Can you please add e2e test? I think we can add test to the existing e2e test app:
1. test app - https://github.com/dapr/dapr/blob/master/tests/apps/hellodapr/app.go
2. test driver - https://github.com/dapr/dapr/blob/master/tests/e2e/hellodapr/hellodapr_test.go

File: pkg/injector/pod_patch.go (hunk 7 above)
Comment (mukundansundar) [reply pkg/injector/pod_patch.go:167]: This code will inject the env vars into all defined/injected containers.

File: pkg/injector/pod_patch.go [hunk 11]
```diff
@@ -142,10 +147,77 @@ func (i *injector) getPodPatchOperations(ar *v1beta1.AdmissionReview,
...
+				break
+			}
+		}
+		if !hasEnv {
+			toAdd = append(toAdd, env)
+		}
+	}
+
+	patchOps := make([]PatchOperation, len(toAdd))
+	for i, value := range toAdd {
```
Comment (youngbupark) [inline pkg/injector/pod_patch.go:211]: We do not need this for-loop if we can use append EnvVar to patchOps at line 206.

Comment (youngbupark) [reply pkg/injector/pod_patch.go:211]: e.g.
```
        var patchOps []PatchOperation
	LoopEnv: for _, env := range addEnv {
		for _, actual := range envs {
			if actual.Name == env.Name {
				hasEnv = true
				goto LoopEnv
			}
		}
                patchOps[i] = PatchOperation{
			Op:    "add",
			Path:  path,
			Value: []corev1.EnvVar{value},
		}
	}
```

Comment (youngbupark) [reply pkg/injector/pod_patch.go:211]: does PathOperation accept the list of values? is it accepting only one value

File: pkg/injector/pod_patch.go [hunk 12]
```diff
@@ -142,10 +147,77 @@ func (i *injector) getPodPatchOperations(ar *v1beta1.AdmissionReview,
...
+		{
+			Name:  userContainerDaprGRPCPortName,
+			Value: strconv.Itoa(sidecarAPIGRPCPort),
+		},
+	}
+	envPatchOps := []PatchOperation{}
+	for i, container := range containers {
+		path := fmt.Sprintf("%s/%d/env", containersPath, i)
+		patchOps := getEnvPatchOperations(container.Env, portEnv, path)
+		if len(patchOps) != 0 {
```
Comment (pruthvidhodda) [inline pkg/injector/pod_patch.go:172]: We don't need this additional check, append will take care of empty slice

**Pull request merged**